### ISSUE TITLE
Iceberg Integration

### ISSUE THREAD

**Issue by linhr** [lakehq/sail#172]
https://github.com/apache/iceberg-rust

**Comment by shehabgamin** [comment]
https://iceberg.apache.org/docs/latest/spark-getting-started/

**Comment by fafacao86** [comment]
> We plan to maintain our in-house Iceberg implementation based on existing libraries such as https://github.com/apache/iceberg-rust.
> 
@linhr  I'm curious why  Lake Sail plans to maintain an in-house iceberg implementation, could you please share some motivation? I'm trying to build an OLAP database on top of iceberg, and I find it hard to extend iceberg often times.

**Comment by linhr** [comment]
> I'm curious why Lake Sail plans to maintain an in-house iceberg implementation, could you please share some motivation? I'm trying to build an OLAP database on top of iceberg, and I find it hard to extend iceberg often times.

We are working on deep integration of Iceberg and other lakehouse formats into Sail's query planning and distributed execution mechanisms. We found it's currently best to maintain our own Iceberg implementation as the existing ones lack integration points to the internals for our needs. That being said, we are still part of the `iceberg-rust` community and we've been involved in a few design discussions. We do hope that the official implementation would become more feature-complete and modular over time, so that it can provide shared building blocks for downstream projects like Sail.

**Comment by fafacao86** [comment]
> > I'm curious why Lake Sail plans to maintain an in-house iceberg implementation, could you please share some motivation? I'm trying to build an OLAP database on top of iceberg, and I find it hard to extend iceberg often times.
> 
> We are working on deep integration of Iceberg and other lakehouse formats into Sail's query planning and distributed execution mechanisms. We found it's currently best to maintain our own Iceberg implementation as the existing ones lack integration points to the internals for our needs. That being said, we are still part of the `iceberg-rust` community and we've been involved in a few design discussions. We do hope that the official implementation would become more feature-complete and modular over time, so that it can provide shared building blocks for downstream projects like Sail.

Got it, thanks for your reply!

**Comment by atakanyenel** [comment]
I'm testing sail to read from iceberg-on-hms tables. I built from https://github.com/lakehq/sail/pull/1929 commit and reading iceberg tables work. But **writing doesn't.** Especially *writeTo/createTable* functions are failing. What is the timeline for te write ability for iceberg tables ?

**Comment by linhr** [comment]
> I'm testing sail to read from iceberg-on-hms tables. I built from [#1929](https://github.com/lakehq/sail/pull/1929) commit and reading iceberg tables work. But **writing doesn't.** Especially _writeTo/createTable_ functions are failing. What is the timeline for te write ability for iceberg tables ?

Thanks for trying out Sail! We have read and write support for Iceberg tables when specifying the file path directly, but there may be a few missing pieces when accessing Iceberg tables in catalogs. Feel free to create a separate issue with more information around your setup. We'd be happy to look into this and provide a better estimate of when those features will be available.

**Comment by atakanyenel** [comment]
I created 2 issues for the problems I have faced:
- https://github.com/lakehq/sail/issues/2056
- https://github.com/lakehq/sail/issues/2055

**Comment by santosh-d3vpl3x** [comment]
Indeed, iceberg support for HMS is incomplete at this moment. Anything HMS requires quite a bit of plumbing and quirky handling on client side (I.e. on sail side). Iceberg is quite a bit more plumbing than what is typical even for that setup so it [was left out of initial iterations](https://docs.lakesail.com/sail/latest/guide/catalog/hms.html#:~:text=Iceberg%2Din%2DHMS%20behavior). I have a couple other high prio things to get out of the door. I will come back to iceberg support later if that is okay until then.

**Comment by linhr** [comment]
> Indeed, iceberg support for HMS is incomplete at this moment. Anything HMS requires quite a bit of plumbing and quirky handling on client side (I.e. on sail side). Iceberg is quite a bit more plumbing than what is typical even for that setup so it [was left out of initial iterations](https://docs.lakesail.com/sail/latest/guide/catalog/hms.html#:~:text=Iceberg%2Din%2DHMS%20behavior). I have a couple other high prio things to get out of the door. I will come back to iceberg support later if that is okay until then.

Thanks for chiming in! @santosh-d3vpl3x Indeed it's a complex interaction between lakehouse formats and catalogs. We're also doing some design work internally to make things more modularized for such deep integration. So hopefully things will become easier in the coming weeks, which might be a good time to collaborate on the feature development and testing.

**Comment by Animeshz** [comment]
Hey, I was trying to experiement with apache polaris as [iceberg catalog](https://docs.lakesail.com/sail/latest/guide/catalog/iceberg-rest.html). I found it really weird that I have to provide `oauth_access_token` instead of oauth client-id & secret. And the token expires in 1 hour.

Any better workaround apart from generating oauth tokens manually?

**Comment by linhr** [comment]
> Hey, I was trying to experiement with apache polaris as [iceberg catalog](https://docs.lakesail.com/sail/latest/guide/catalog/iceberg-rest.html). I found it really weird that I have to provide `oauth_access_token` instead of oauth client-id & secret. And the token expires in 1 hour.
> 
> Any better workaround apart from generating oauth tokens manually?

Thanks for trying out the Iceberg REST catalog in Sail!

The `oauth_access_token` option was actually a defect in our current design, and we plan to deprecate that in the future. We should either accept a bearer token or handle the full OAuth flow properly. The OAuth flow would be similar to how we fetch/refresh tokens for Unity Catalog and OneLake catalog. This is not implemented yet, but contributions are welcome!
